# Module 3: Influence Measures and Network Centralization


*The notes in this module use NetworkX version 1.11. I haven't tested any of the code on NetworkX version 2.*

## About this Module

In Module Three, you'll explore ways of measuring the importance or centrality of a node in a network, using measures such as Degree, Closeness, and Betweenness centrality, Page Rank, and Hubs and Authorities. You'll learn about the assumptions each measure makes, the algorithms we can use to compute them, and the different functions available on NetworkX to measure centrality. In the assignment, you'll practice choosing the most appropriate centrality measure on a real-world setting.

## Learning objectives

* Identify and define several network centrality measures.
* Describe the differences and similarities between several centrality measures.
* Measure the centrality of nodes in a network using NetworkX.
* Describe how network centrality measures can be used for real world applications.
* Apply centrality analysis to real world networks.

## Lecture 1: Degree and Closeness Centrality

### Network Centrality

Centrality measures identify the most importat nodes in the network:

* Influential nodes in a social network.
* Nodes that disseminate information to other nodes.
* Hubs in a transportation network.
* Important pages on the web.
* Nodes that prevent the network from breaking up.

List of centrality measures that are commonly used:

* **Degree Centrality**    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  <-- We cover this one
* **Closeness Centrality**  &nbsp;&nbsp;&nbsp;   <-- We cover this one
* Betweenness Centrality
* Load Centrality
* Page Rank
* Katz Centrality
* Percolation Centrality
    

### Degree Centrality

**Assumption**: Important nodes have mane connections.

This is the most basic way you can define centrality: number of neighbors.

* For undirected networks: Use **degree**.

* For directed networks: Use **in-degree** or **out-degree**.


Mathematical definition of Degree Centrality of a node $v$:

$$ C_{deg}(v) = \frac{d_v}{ \left\vert N \right\vert - 1} $$ 

where:

* N: Set of nodesin the network,
* $d_v$: degree of the node $v$.

So, by this definition: 

* a node that is connected to all nodes has **centrality 1**, 
* a node that is not connected to any node has **centrality 0**.

In [3]:
import networkx as nx

# Let's load the karate club data
G = nx.karate_club_graph()
G = nx.convert_node_labels_to_integers(G, first_label=1)

We can use the function `nx.degree_centrality(G)` to measure the centrality of all the nodes in the network:


In [5]:
degCent = nx.degree_centrality(G)
degCent  # returns a dict {node->centrality}

{1: 0.48484848484848486,
 2: 0.2727272727272727,
 3: 0.30303030303030304,
 4: 0.18181818181818182,
 5: 0.09090909090909091,
 6: 0.12121212121212122,
 7: 0.12121212121212122,
 8: 0.12121212121212122,
 9: 0.15151515151515152,
 11: 0.09090909090909091,
 12: 0.030303030303030304,
 13: 0.06060606060606061,
 14: 0.15151515151515152,
 18: 0.06060606060606061,
 20: 0.09090909090909091,
 22: 0.06060606060606061,
 32: 0.18181818181818182,
 31: 0.12121212121212122,
 10: 0.06060606060606061,
 28: 0.12121212121212122,
 29: 0.09090909090909091,
 33: 0.36363636363636365,
 17: 0.06060606060606061,
 34: 0.5151515151515151,
 15: 0.06060606060606061,
 16: 0.06060606060606061,
 19: 0.06060606060606061,
 21: 0.06060606060606061,
 23: 0.06060606060606061,
 24: 0.15151515151515152,
 26: 0.09090909090909091,
 30: 0.12121212121212122,
 25: 0.09090909090909091,
 27: 0.06060606060606061}

We can look at the centrality of node 34 *(see lecture video at time 4:55)*

We can see its centrality is 0.515, which is the result of dividing its 17 connection over 34 - 1 (total nodes minus 1):

In [6]:
degCent[34]

0.5151515151515151

#### Degree Centrality for directed networks

For directed networks we use the same formula but we have to decide if how we count the connections: in-connections or out-connections.

* In-degree Centrality: 
$$ C_{indeg}(v) = \frac{d^{in}_v}{ \left\vert N \right\vert - 1} $$

* Out-degree Centrality: 
$$ C_{outdeg}(v) = \frac{d^{out}_v}{ \left\vert N \right\vert - 1} $$

In [7]:
# directed graph used as example 
G = nx.DiGraph()
G.add_edges_from([('A', 'B'), ('C', 'A'), ('A', 'E'), ('G', 'A'), ('A', 'N'), 
                  ('B', 'C'), ('D', 'B'), ('B', 'E'), ('C', 'D'), ('E', 'C'), 
                  ('D', 'E'), ('E', 'D'), ('F', 'G'), ('I', 'F'), ('J', 'F'), 
                  ('H', 'G'), ('I', 'G'), ('G', 'J'), ('I', 'H'), ('H', 'I'), 
                  ('I', 'J'), ('J', 'O'), ('O', 'J'), ('K', 'M'), ('K', 'L'), 
                  ('O', 'K'), ('O', 'L'), ('N', 'L'), ('L', 'M'), ('N', 'O')])

In [11]:
G.number_of_nodes(), G.number_of_edges()

(15, 30)

##### In-degree calculation

We can calculate the in-degree centrality for all the nodes with the fucntion `nx.in_degree_centrality(G)`. Then we can see that for node A the centrality is 2/14 *(see lecture video at time 5:43)*:

In [8]:
indegCent = nx.in_degree_centrality(G)
indegCent['A']  # 2/14 see lecture video at time 5:43

0.14285714285714285

In [12]:
indegCent['L']  # 3/14 see lecture video at time 5:43

0.21428571428571427

##### Out-degree calculation

We can calculate the out-degree centrality for all the nodes with the fucntion `nx.out_degree_centrality(G)`. Then we can see that for node A the centrality is 3/14 *(see lecture video at time 6:15)*:

In [15]:
outdegCent = nx.out_degree_centrality(G)
outdegCent['A']  # 3/14 see lecture video at time 6:15

0.21428571428571427

In [16]:
outdegCent['L']  # 1/14 see lecture video at time 6:15

0.07142857142857142

### Closeness Centrality

#### Closeness Centrality for connected graphs

**Assumption**: Important nodes are a short distance away from all other nodes in the network. 


Mathematical definition of Closeness Centrality of a node $v$:

$$ C_{close}(v) = \frac{ \left\vert N \right\vert - 1}{\sum_{u \in N} d(v,u)} $$ 

where:

* N: Set of nodes in the network,
* $d(v,u)$: length of the shortest path from $v$ to $u$. (a.k.a. distance from v to u).


In [17]:
# Let's load the karate club data
G = nx.karate_club_graph()
G = nx.convert_node_labels_to_integers(G, first_label=1)

We can calculate the closeness centrality of all the node with `nx.closeness_centrality(G)`

In [19]:
closeCent = nx.closeness_centrality(G)
closeCent  # dict {node-> cent}

{1: 0.5689655172413793,
 2: 0.4852941176470588,
 3: 0.559322033898305,
 4: 0.4647887323943662,
 5: 0.3793103448275862,
 6: 0.38372093023255816,
 7: 0.38372093023255816,
 8: 0.44,
 9: 0.515625,
 11: 0.3793103448275862,
 12: 0.36666666666666664,
 13: 0.3707865168539326,
 14: 0.515625,
 18: 0.375,
 20: 0.5,
 22: 0.375,
 32: 0.5409836065573771,
 31: 0.4583333333333333,
 10: 0.4342105263157895,
 28: 0.4583333333333333,
 29: 0.4520547945205479,
 33: 0.515625,
 17: 0.28448275862068967,
 34: 0.55,
 15: 0.3707865168539326,
 16: 0.3707865168539326,
 19: 0.3707865168539326,
 21: 0.3707865168539326,
 23: 0.3707865168539326,
 24: 0.39285714285714285,
 26: 0.375,
 30: 0.38372093023255816,
 25: 0.375,
 27: 0.3626373626373626}

In [20]:
closeCent[32]

0.5409836065573771

Let's see how we obtained this result for node 32. Here we calculate the sum of all the distances from node 32 to all the other nodes:

In [24]:
sum(nx.shortest_path_length(G, 32).values())

61

In [26]:
# should give the same than closeCent[32]
(len(G.nodes()) - 1) / 61

0.5409836065573771

#### Closeness Centrality for disconnected graphs (General case)

The previous definition of Closeness Centrality does not really work when some node can not reach all other nodes. In particular for directed graphs it is often the case that there are multiple connected components, so the graph is disconnected. 

So, how can we extend the definition for the general case of a disconnected graph?

##### Option 1: Consider only nodes L can reach:

$$ C_{close}(L) = \frac{ \left\vert R(L) \right\vert }{\sum_{u \in R(L)} d(L,u)} $$ 

where:
* $L$: source node 
* $R(L)$: Set of nodes L can reach in the network,
* $d(L,u)$: length of the shortest path from $L$ to $u$. (a.k.a. distance from L to u).

**Problem**: For a node that only is connected to one node: 
$ C_{close}(L) = \frac{1}{1} $ , this is the maximum centrality a node can have, so this is not appropriate for a node that only has one connection.

##### Option 2: Consider only nodes L can reach with normalization

Consider only nodes L can reach and normalize by the fraction of nodes L can reach:

$$ C_{close}(L) = \left[ \frac{\left\vert R(L) \right\vert}{\left\vert N \right\vert - 1}  \right] \frac{ \left\vert R(L) \right\vert }{\sum_{u \in R(L)} d(L,u)} $$ 

where:
* $L$: source node 
* $R(L)$: Set of nodes L can reach in the network,
* $d(L,u)$: length of the shortest path from $L$ to $u$. (a.k.a. distance from L to u).


**Problem solved**: So using this option nodes with only one connection get small values of centrality, which is more reasonable.

**Notice**: Note that when the graph is connected this definition matches the original definition for connected graphs because $R(L) = N - 1$

#### Measure Closeness Centrality with networkx



In [28]:
# directed graph used as example 
G = nx.DiGraph()
G.add_edges_from([('A', 'B'), ('C', 'A'), ('A', 'E'), ('G', 'A'), ('A', 'N'), 
                  ('B', 'C'), ('D', 'B'), ('B', 'E'), ('C', 'D'), ('E', 'C'), 
                  ('D', 'E'), ('E', 'D'), ('F', 'G'), ('I', 'F'), ('J', 'F'), 
                  ('H', 'G'), ('I', 'G'), ('G', 'J'), ('I', 'H'), ('H', 'I'), 
                  ('I', 'J'), ('J', 'O'), ('O', 'J'), ('K', 'M'), ('K', 'L'), 
                  ('O', 'K'), ('O', 'L'), ('N', 'L'), ('L', 'M'), ('N', 'O')])

For option 1 (no normalization): *(see lecture video time 11:20)*

In [30]:
closeCent = nx.closeness_centrality(G, normalized=False)
closeCent['L']

1.0

For option 2 (with normalization): *(see lecture video time 11:25)*

In [31]:
closeCent = nx.closeness_centrality(G, normalized=True)
closeCent['L']

0.07142857142857142

### Summary

* Degree Centrality
    - Assumption: important nodes have many connections.
    - `nx.degree_centrality(G)` : for undirected graphs
    - `nx.in_degree_centrality(G)`  : for directed graphs
    - `nx.out_degree_centrality(G)` : for directed graphs


* Closeness Centrality:
    - Assumption: important nodes have short distances to all other nodes.
    - `nx.closeness_centrality(G, normalize=True)` : for undirected graphs or directed graphs


## Lecture 2: Betweenness Centrality

**Assumption**: Important nodes are those who connect other nodes.


$$ C_{btw}(v) = \sum_{s,t \in N} \frac {\sigma_{s,t}(v)}{\sigma_{s,t}} $$ 

where:
* $\sigma_{s,t}$: the number of shortest paths between nodes s and t. 
* $\sigma_{s,t}(v)$: the number of shortest paths between nodes s and t *that pass through node v*.


**Option to include endpoints**: You can exclude or include node $v$ as nodes $s, t$ in the computation of $C_{btw}(v)$

Networkx will let you make this choice in their functions. 

**Problem**: What if not all nodes can reach each other (disconnected graph)? This happens often in directed graphs.

Note that if 2 nodes $a, b$ can not reach each other: $\sigma_{a,b}=0$ , which makes the the formula undefined.

How to perform the computation in this case?

**Solution**: When computing betweenness centrality, we only consider nodes $s, t$ such that there is at least one path between them. 

#### Normalization in Betweenness Centrality

The values are larger for graphs with many nodes. 

To control for this we divide centrality values by the numbrer of pair of nodes in the graph (excluding $v$):

* $ \frac{1}{2} ( \vert N \vert - 1) ( \vert N \vert - 2) $ : for undirected graphs


* $ ( \vert N \vert - 1) ( \vert N \vert - 2) $  : for directed graphs

### Betweenness Centrality in networkx

In [32]:
# Let's load the karate club data
G = nx.karate_club_graph()
G = nx.convert_node_labels_to_integers(G, first_label=1)

We can calculate the closeness centrality of all the node with `nx.betweenness_centrality(G)`

In [35]:
btwnCent = nx.betweenness_centrality(G, normalized=True, endpoints=False)
btwnCent

{1: 0.4376352813852815,
 2: 0.05393668831168831,
 3: 0.14365680615680615,
 4: 0.011909271284271283,
 5: 0.0006313131313131313,
 6: 0.02998737373737374,
 7: 0.029987373737373736,
 8: 0.0,
 9: 0.05592682780182782,
 11: 0.0006313131313131313,
 12: 0.0,
 13: 0.0,
 14: 0.045863395863395856,
 18: 0.0,
 20: 0.03247504810004811,
 22: 0.0,
 32: 0.13827561327561327,
 31: 0.014411976911976905,
 10: 0.0008477633477633478,
 28: 0.022333453583453587,
 29: 0.0017947330447330447,
 33: 0.14524711399711404,
 17: 0.0,
 34: 0.30407497594997596,
 15: 0.0,
 16: 0.0,
 19: 0.0,
 21: 0.0,
 23: 0.0,
 24: 0.017613636363636363,
 26: 0.0038404882154882162,
 30: 0.0029220779220779218,
 25: 0.0022095959595959595,
 27: 0.0}

Let's filter the above result to get the top 5 nodes with the highest betweenness centrality scores *(see lecture video time 10:20 for a figure showing the positions of these nodes in the Karate graph)*:

In [37]:
from operator import itemgetter

sorted(btwnCent.items(), key=itemgetter(1), reverse=True)[0:5]

[(1, 0.4376352813852815),
 (34, 0.30407497594997596),
 (33, 0.14524711399711404),
 (3, 0.14365680615680615),
 (32, 0.13827561327561327)]

#### Betweenness Centrality - Computational complexity - k parameter

This calculation is very computationaly expensive. Depending on the algorithm it can take $O({\vert N \vert}^3)$ time.

**Approximation**: Limit number of pair of nodes $s, t$ to be considered.

In networkx you can achieve this by passing the parameter k:

In [38]:
btwnCent = nx.betweenness_centrality(G, normalized=True, endpoints=False, k=10)
btwnCent

{1: 0.4273151154401154,
 2: 0.05354662698412698,
 3: 0.11270472582972581,
 4: 0.0070296717171717165,
 5: 0.0010732323232323232,
 6: 0.015561868686868688,
 7: 0.014488636363636363,
 8: 0.0,
 9: 0.06956845238095238,
 11: 0.0,
 12: 0.0,
 13: 0.0,
 14: 0.06237779581529581,
 18: 0.0,
 20: 0.05172213203463203,
 22: 0.0,
 32: 0.11955041486291486,
 31: 0.028134018759018754,
 10: 0.0006746031746031745,
 28: 0.010762987012987012,
 29: 0.0006746031746031745,
 33: 0.16817550505050508,
 17: 0.0,
 34: 0.3942288961038961,
 15: 0.0,
 16: 0.0,
 19: 0.0,
 21: 0.0,
 23: 0.0,
 24: 0.008049242424242424,
 26: 0.0012878787878787877,
 30: 0.004967532467532467,
 25: 0.0,
 27: 0.0}

If we look now at the top 5 nodes with the highest betweenness centrality scores, the results can be a bit different, but hopefully close enough:

In [39]:
sorted(btwnCent.items(), key=itemgetter(1), reverse=True)[0:5]

[(1, 0.4273151154401154),
 (34, 0.3942288961038961),
 (33, 0.16817550505050508),
 (32, 0.11955041486291486),
 (3, 0.11270472582972581)]

#### Betweenness Centrality - Subsets

Sometimes you want to be able to select the sets of source nodes and target nodes to be used when computing the Betweenness Centrality. Maybe because you are interested in studying the nodes that are important for the comunication between 2 specific sections of the network.

In networkx you can do this by using the fucntion `nx.betweenness_centrality_subset(G, source_set, target_set)`:

In [41]:
sources = [34, 33, 21, 30, 16, 27, 15, 23, 10]
targets = [1, 4, 13, 11, 6, 12, 17, 7]

btwnCent_subset = nx.betweenness_centrality_subset(G, sources, targets, normalized=True)
btwnCent_subset

{1: 0.04899515993265994,
 2: 0.0,
 3: 0.018368205868205867,
 4: 0.0021412037037037038,
 5: 0.0,
 6: 0.004261363636363636,
 7: 0.004261363636363636,
 8: 0.0,
 9: 0.014519450456950456,
 11: 0.0,
 12: 0.0,
 13: 0.0,
 14: 0.012970478595478596,
 18: 0.0,
 20: 0.007804232804232804,
 22: 0.0,
 32: 0.014519450456950456,
 31: 0.0,
 10: 0.0,
 28: 0.0,
 29: 0.0,
 33: 0.01664712602212602,
 17: 0.0,
 34: 0.028807419432419434,
 15: 0.0,
 16: 0.0,
 19: 0.0,
 21: 0.0,
 23: 0.0,
 24: 0.0,
 26: 0.0,
 30: 0.0,
 25: 0.0,
 27: 0.0}

Let's filter the above result to get the top 5 nodes with the highest betweenness centrality scores *(see lecture video time 13:10 for a figure showing the positions of  of the source and target nodes in the Karate graph)*:

In [43]:
sorted(btwnCent_subset.items(), key=itemgetter(1), reverse=True)[0:5]

[(1, 0.04899515993265994),
 (34, 0.028807419432419434),
 (3, 0.018368205868205867),
 (33, 0.01664712602212602),
 (9, 0.014519450456950456)]

#### Betweenness Centrality - Edges

We can use betweenness centrality to find the important edges instead of nodes, using the same formula:

$$ C_{btw}(e) = \sum_{s,t \in N} \frac {\sigma_{s,t}(e)}{\sigma_{s,t}} $$ 

where:
* $\sigma_{s,t}$: the number of shortest paths between nodes s and t. 
* $\sigma_{s,t}(v)$: the number of shortest paths between nodes s and t *that pass through edge e*.


We can calculate the edge betweenness centrality with networkx function `nx.edge_betweenness_centrality`:

In [45]:
btwnCent_edge = nx.edge_betweenness_centrality(G, normalized=True)
btwnCent_edge

{(1, 2): 0.025252525252525245,
 (1, 3): 0.07778768072885717,
 (1, 4): 0.02049910873440285,
 (1, 5): 0.0522875816993464,
 (1, 6): 0.07813428401663694,
 (1, 7): 0.07813428401663695,
 (1, 8): 0.0228206434088787,
 (1, 9): 0.07423959482783016,
 (1, 11): 0.0522875816993464,
 (1, 12): 0.058823529411764705,
 (1, 13): 0.04652406417112298,
 (1, 14): 0.042371898254251195,
 (1, 18): 0.04012392835922249,
 (1, 20): 0.045936960642843,
 (1, 22): 0.04012392835922249,
 (1, 32): 0.12725999490705373,
 (2, 3): 0.023232323232323233,
 (2, 4): 0.0077243018419489,
 (2, 8): 0.007422969187675069,
 (2, 14): 0.01240556828792123,
 (2, 18): 0.018699601052542224,
 (2, 20): 0.014633732280791106,
 (2, 22): 0.018699601052542224,
 (2, 31): 0.032280791104320514,
 (3, 4): 0.022430184194890075,
 (3, 8): 0.025214328155504617,
 (3, 9): 0.009175791528732703,
 (3, 10): 0.030803836686189627,
 (3, 14): 0.007630931160342923,
 (3, 28): 0.041192032368502954,
 (3, 29): 0.022782446311858072,
 (3, 33): 0.06898678663384546,
 (4, 8): 0.0

Let's filter the output to find the top 5 edges with the highest betweenness centrality:

In [46]:
sorted(btwnCent_edge.items(), key=itemgetter(1), reverse=True)[0:5]

[((1, 32), 0.12725999490705373),
 ((1, 7), 0.07813428401663695),
 ((1, 6), 0.07813428401663694),
 ((1, 3), 0.07778768072885717),
 ((1, 9), 0.07423959482783016)]

You can select the source and target subsets to calculate the betweenness centrality for the edges too (similar to the nodes).

For that you use networkx function `nx.edge_betweenness_centrality_subset` *(see lecture video time 15:55)*:

In [47]:
sources = [34, 33, 21, 30, 16, 27, 15, 23, 10]
targets = [1, 4, 13, 11, 6, 12, 17, 7]

btwnCent_edge_subset = nx.edge_betweenness_centrality_subset(G, sources, targets, normalized=True)
btwnCent_edge_subset

{(1, 2): 0.0,
 (1, 3): 0.01211343123107829,
 (1, 4): 0.0,
 (1, 5): 0.0,
 (1, 6): 0.012032085561497326,
 (1, 7): 0.012032085561497326,
 (1, 8): 0.0,
 (1, 9): 0.01366536513595337,
 (1, 11): 0.00802139037433155,
 (1, 12): 0.00802139037433155,
 (1, 13): 0.006006139829669241,
 (1, 14): 0.007345160286336755,
 (1, 18): 0.0,
 (1, 20): 0.007345160286336755,
 (1, 22): 0.0,
 (1, 32): 0.01366536513595337,
 (2, 3): 0.0,
 (2, 4): 0.0,
 (2, 8): 0.0,
 (2, 14): 0.0,
 (2, 18): 0.0,
 (2, 20): 0.0,
 (2, 22): 0.0,
 (2, 31): 0.0,
 (3, 4): 0.005174291938997821,
 (3, 8): 0.0,
 (3, 9): 0.0,
 (3, 10): 0.0071301247771836,
 (3, 14): 0.0,
 (3, 28): 0.0,
 (3, 29): 0.0,
 (3, 33): 0.010157598392892509,
 (4, 8): 0.0,
 (4, 13): 0.0020152505446623093,
 (4, 14): 0.004862348979996039,
 (5, 7): 0.0,
 (5, 11): 0.0,
 (6, 7): 0.0,
 (6, 11): 0.0,
 (6, 17): 0.004010695187165775,
 (7, 17): 0.004010695187165775,
 (9, 31): 0.0,
 (9, 33): 0.006320204849616614,
 (9, 34): 0.007345160286336755,
 (14, 34): 0.012207509266332794,
 (20, 3

Let's filter the output to find the top 5 edges with the highest betweenness centrality between the 2 set of source and target nodes:

In [48]:
sorted(btwnCent_edge_subset.items(), key=itemgetter(1), reverse=True)[0:5]

[((1, 9), 0.01366536513595337),
 ((1, 32), 0.01366536513595337),
 ((14, 34), 0.012207509266332794),
 ((1, 3), 0.01211343123107829),
 ((1, 6), 0.012032085561497326)]

### Sumary



* **Betweenness centrality assumption**: important nodes connect other nodes.


* **Normalization**: Divide by the number of pair of nodes.


* **Approximation**: Limit number of pairs $(s, t)$ (paramenter k).


* **Subsets**: You can define which subsets to use as source and target to compute Betweenness centrality.


* **Edge betweenness centrality**: We can use the same framework to find important edges.

## Lecture 3: Basic PageRank

**Assumption**: Important nodes are those who have many in-links from other important nodes.

PageRank can be used in any type of network, but it is most useful for directed networks.



**PageRank is an iterative algorithm**, it takes several steps to converge to a solution:

n = number of nodes in the network.

k = number of steps.

1. Assign all nodes a PageRank of 1/n.


2. Perform the *Basic PageRank Update Rule* k times: 
    * **Basic PageRank Update Rule**: Each node gives an equal share of its current PageRank to each node it links to.
    * The new iteration PageRank of a node will be the sum of all PageRank it received from other nodes.
    

For most networks PageRank values converge as k gets larger ($k \to \infty$) 

## Lecture 4: Scaled PageRank

* The basic PageRank of a node can be interpreted as the probability that a random walk lands on the node after k steps.


* Basic PageRank has the problem that, in some networks, a few nodes can "suck up" all the PageRank from the network. *(See lecture video time 2:39 for a simple example)*


* To fix this problem **Scaled PageRank** introduces a parameter $\alpha$, such that the random walker can choose to jump to a random node (even not connected nodes) with probability $p = 1 - \alpha$


* Typically $\alpha$ is between 0.8 and 0.9. 


* The particular value of PageRank will depend on the chosen $\alpha$.


* For most networks Scaled PageRank will converge when k gets larger ($k \to \infty$).


* To compute Scaled PageRank you can use NetworkX function `nx.pagerank(G, alpha=0.8)`

## Lecture 5: Hubs and Authorities

#### Hyperlink-Induced Topic Search (HITS; also known as hubs and authorities)


Given a query to a search engine:

* **Root set**: Set of highly relevant web pages (e.g. pages that contain the query string) - *potential authorities*.


* Find all pages that link to a page in Root - *potential hubs*.


* **Base set**: root node plus any node that links to a node in Root.


* Consider all edges that conecting nodes in the base set.


*(See lecture video time 1:42 for a figure representing the root and base sets on a network)*

### HITS Algorithm

Compute k iterations of the HITS algorithm to assign the *authority score* and the *hub score* to each node:


1. Assign to each node an authority and hub score of 1.


2. Apply the **Authority Update Rule**: each node's **authority** score is the sum of hub scores of each node that **points to it** (in-links).


3. Apply the **Hub Update Rule**: each node's **hub** score is the sum of authority scores of each node that **it points to** (out-links).


4. **Normalize** Authority and Hub scores: $ auth(j) = \frac{ auth(j) }{ \sum_{i \in N}{auth(i)}} $ 


5. Repeat the process $k$ times.

*(See Lecture video at time 4:00 for an example computation of the HITS algorithm in a simple network)*

#### HITS with networkx

You can use NetworkX function `nx.hits(G)` to compute the hub and authority.

This function output is 2 dicts, keyed by node, with the hub and authority scores for the nodes.

### Summary

* The HITS algorithm starts by constructing a *root set* of relevant web pages and extending it to a *base set*.


* Nodes that have incoming edges from *good hubs* are *good authorities*, and nodes that have outgoing edges to *good authorities* are *good hubs*.


* For most networks the authority and hub scores converge when k gets larger ($k \to \infty$).


* You can use NetworkX function `nx.hits(G)` to compute the hub and authority.

## Lecture 6: Centrality examples

The lecture just present a comparison of the different Centrality measure scores calculated for an example network.

**Conclusions**:

* Different centrality measures produce different node ranking, but there are some commonalities.


* Centrality measures make different assumptions about what it means to be a "central" node.


* The best centrality measure depend on the context of the network one is analyzing.

--------------------

## Extra information

* NetworkX official documentation on Centrality has a nice review of the centrality measures and even include some figures and formulas from the lecures: https://pynetwork.readthedocs.io/en/latest/influence_central.html